# Mini-Projeto Avaliativo - Análise de Dados no Varejo
**Módulo 1 - Análise de Dados com Python (Turma: Analise_de_Dados_T3)**

### Objetivos
1. **ETL & Limpeza**: Carregar a base, tratar nulos, duplicatas e ajustar tipos de dados.
2. **Estatística Descritiva**: Mapear a distribuição da coluna de número de filhos dos clientes.
3. **Agrupamentos**: Identificar padrões com `groupby()`.
4. **Insights**: Reportar as conclusões da análise.

--- 
## Sprint 1: Importação e Diagnóstico Inicial dos Dados

In [1]:
# Bibliotecas

import pandas as pd
from datetime import datetime

In [2]:
# Leitura do CSV ajustando o separador para ;
df = pd.read_csv(
    "Base Varejo.csv", sep=";"
)

In [3]:
# Primeiras linhas
df.head()

,DATA,CO_ID,CL_ID,CL_GENERO,CL_EC,CL_FHL,CL_SEG,PR_ID,PR_CAT,PR_NOME,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13
0,01/02/2019,1000,534,M,4,1,C,67,BEBIDAS,REFRIGERANTE GUARANA,NaN,NaN,NaN,NaN
1,01/02/2019,1000,534,M,4,1,C,70,BEBIDAS,REFRIGERANTE OUTROS,NaN,NaN,NaN,NaN
2,01/02/2019,1000,534,M,4,1,C,178,HIGIENE,LENCO UMEDECIDO,NaN,NaN,NaN,NaN
3,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI,NaN,NaN,NaN,NaN
4,01/02/2019,1000,534,M,4,1,C,175,LIMPEZA,LIMPADOR MULTIUSO,NaN,NaN,NaN,NaN


In [4]:
# Quantidade de linhas e colunas
df.shape

(830000, 14)

In [5]:
# Informações
print("Informações de Tipos de Dados")
print(df.info())

Informações de Tipos de Dados
<class 'pandas.DataFrame'>
RangeIndex: 830000 entries, 0 to 829999
Data columns (total 14 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   DATA         830000 non-null  str    
 1   CO_ID        830000 non-null  int64  
 2   CL_ID        830000 non-null  int64  
 3   CL_GENERO    830000 non-null  str    
 4   CL_EC        830000 non-null  int64  
 5   CL_FHL       830000 non-null  int64  
 6   CL_SEG       830000 non-null  str    
 7   PR_ID        830000 non-null  int64  
 8   PR_CAT       830000 non-null  str    
 9   PR_NOME      830000 non-null  str    
 10  Unnamed: 10  0 non-null       float64
 11  Unnamed: 11  0 non-null       float64
 12  Unnamed: 12  0 non-null       float64
 13  Unnamed: 13  0 non-null       float64
dtypes: float64(4), int64(5), str(5)
memory usage: 88.7 MB
None


In [6]:
print("--- Contagem de Nulos por Coluna ---")
print(df.isna().sum())

--- Contagem de Nulos por Coluna ---
DATA                0
CO_ID               0
CL_ID               0
CL_GENERO           0
CL_EC               0
CL_FHL              0
CL_SEG              0
PR_ID               0
PR_CAT              0
PR_NOME             0
Unnamed: 10    830000
Unnamed: 11    830000
Unnamed: 12    830000
Unnamed: 13    830000
dtype: int64


In [7]:
print(f"Total de Linhas Duplicadas Inteiras: {df.duplicated().sum()}")

Total de Linhas Duplicadas Inteiras: 96553


--- 
## Sprint 2 (Transformação de Strings, Integer e Float e Datetime)

In [8]:
# Necessário remover colunas vazias geradas pelos ; do final da linha
df = df.dropna(how="all", axis=1)

df.head()

,DATA,CO_ID,CL_ID,CL_GENERO,CL_EC,CL_FHL,CL_SEG,PR_ID,PR_CAT,PR_NOME
0,01/02/2019,1000,534,M,4,1,C,67,BEBIDAS,REFRIGERANTE GUARANA
1,01/02/2019,1000,534,M,4,1,C,70,BEBIDAS,REFRIGERANTE OUTROS
2,01/02/2019,1000,534,M,4,1,C,178,HIGIENE,LENCO UMEDECIDO
3,01/02/2019,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI
4,01/02/2019,1000,534,M,4,1,C,175,LIMPEZA,LIMPADOR MULTIUSO


In [9]:
# Limpeza de Strings e Remoção de Caracteres Especiais
cols_texto = ['PR_CAT', 'PR_NOME', 'CL_GENERO', 'CL_SEG']
for col in cols_texto:
    df[col] = df[col].astype(str).str.strip()
    df[col] = df[col].str.replace(r'[^\w\s-]', '', regex=True)

In [10]:
# Limpeza de Inteiros
cols_inteiros = ['CO_ID', 'CL_ID', 'CL_EC', 'CL_FHL', 'PR_ID']
for col in cols_inteiros:
    df[col] = pd.to_numeric(df[col].astype(str).str.replace(r'\D', '', regex=True), errors='coerce')

In [11]:
# Conversão da coluna DATA para datetime
df['DATA'] = pd.to_datetime(df['DATA'], dayfirst=True, errors='coerce')

In [12]:
df

,DATA,CO_ID,CL_ID,CL_GENERO,CL_EC,CL_FHL,CL_SEG,PR_ID,PR_CAT,PR_NOME
0,2019-02-01,1000,534,M,4,1,C,67,BEBIDAS,REFRIGERANTE GUARANA
1,2019-02-01,1000,534,M,4,1,C,70,BEBIDAS,REFRIGERANTE OUTROS
2,2019-02-01,1000,534,M,4,1,C,178,HIGIENE,LENCO UMEDECIDO
3,2019-02-01,1000,534,M,4,1,C,4,ALIMENTOS,ABACAXI
4,2019-02-01,1000,534,M,4,1,C,175,LIMPEZA,LIMPADOR MULTIUSO
...,...,...,...,...,...,...,...,...,...,...
829995,2022-08-19,919822,155,F,2,0,B,183,ALIMENTOS,KETCHUP
829996,2022-08-19,919822,155,F,2,0,B,56,ALIMENTOS,QUEIJO MUSSARELA
829997,2022-08-19,919822,155,F,2,0,B,227,ALIMENTOS,ARROZ
829998,2022-08-19,919822,155,F,2,0,B,214,ALIMENTOS,CEBOLA


--- 
## Sprint 3 (Limpeza de Nulos e Duplicatas):

In [13]:
# 1. Preenchimento de categorias ausentes
df['PR_CAT'] = df['PR_CAT'].apply(
    lambda categoria: 'Sem Categoria' if categoria == 'ND' else categoria
)

In [14]:
# 2. Tratamento do número de filhos
df['CL_FHL'] = df['CL_FHL'].apply(
    lambda x: 0 if pd.isna(x) else x
)

--- 
## Sprint 4 (Estatística Descritiva):

In [15]:
# Cálculo das estatísticas descritivas para a coluna de Número de Filhos do Cliente
estatisticas = {
        'Contagem': df["CL_FHL"].count(),
        'Média': df["CL_FHL"].mean(),
        'Mediana': df["CL_FHL"].median(),
        'Desvio Padrão': df["CL_FHL"].std(),
        'Moda': df["CL_FHL"].mode()[0] if not df["CL_FHL"].mode().empty else np.nan,
        'Mínimo': df["CL_FHL"].min(),
        'Máximo': df["CL_FHL"].max(),
        '1º Quartil (25%)': df["CL_FHL"].quantile(0.25),
        '3º Quartil (75%)': df["CL_FHL"].quantile(0.75)
    }
df_estatisticas = pd.DataFrame(list(estatisticas.items()), columns=['Métrica', 'Valor'])
print(df_estatisticas)


            Métrica         Valor
0          Contagem  830000.00000
1             Média       1.14654
2           Mediana       0.00000
3     Desvio Padrão       1.41696
4              Moda       0.00000
5            Mínimo       0.00000
6            Máximo       4.00000
7  1º Quartil (25%)       0.00000
8  3º Quartil (75%)       2.00000


In [16]:
# Agrupamento 1: Total de Itens Vendidos e Compras Únicas por Categoria
vendas_categoria = df.groupby('PR_CAT').agg(
    Qtd_Itens=('PR_ID', 'count'),
    Qtd_Notas_Unicas=('CO_ID', 'nunique')
).reset_index()

print("--- Agrupamento por Categoria ---\n")
print(vendas_categoria)

--- Agrupamento por Categoria ---

          PR_CAT  Qtd_Itens  Qtd_Notas_Unicas
0     ACESSORIOS      14557              9224
1      ALIMENTOS     434767             18267
2        BEBIDAS      43299             14735
3        HIGIENE     155574             17610
4        LIMPEZA     145754             17475
5            PET      32399             13495
6  Sem Categoria       3650              3228


In [17]:
# Agrupamento 2: Vendas por Segmentação Econômica do Cliente
vendas_segmento = df.groupby('CL_SEG').agg(
    Qtd_Itens=('PR_ID', 'count'),
    Qtd_Clientes_Unicos=('CL_ID', 'nunique')
).reset_index()

print("\n--- Agrupamento por Segmento do Cliente ---")
print(vendas_segmento)


--- Agrupamento por Segmento do Cliente ---
  CL_SEG  Qtd_Itens  Qtd_Clientes_Unicos
0      A      67736                   84
1      B     530163                  645
2      C     232101                  271


--- 
## Sprint 5 (Relatório e Documentação):

In [18]:
# Exportação da base limpa para CSV
df.to_csv('df_limpo.csv', index=False)
print("Base de dados tratada salva com sucesso em 'df_limpo.csv'.")

Base de dados tratada salva com sucesso em 'df_limpo.csv'.


### Conclusões e Insights da Análise Exploratória (AED)

1. **Integridade das Vendas:** O histórico completo de 830.000 registros transacionais foi preservado, evitando a exclusão indevida de itens vendidos.
2. **Tratamento de Nulos:** Os registros ausentes na categoria do produto (`PR_CAT`) foram imputados como `'Sem Categoria'`.
3. **Higienização de Estrutura:** As colunas vazias geradas pelo separador de texto `;` foram eliminadas e a coluna `DATA` foi convertida para o tipo `datetime`.
4. **Perfil dos Consumidores:** A análise da variável `CL_FHL` indicou uma média de 1.15 filhos por cliente, com mediana e moda em 0, apontando que a maioria dos clientes cadastrados não possui filhos.
5. **Comportamento por Categoria e Segmento:** Os agrupamentos por `groupby()` identificaram as categorias de produtos mais vendidas e a distribuição do volume de compras de acordo com o segmento socioeconômico do cliente.